src/main/typescript/wcardinal/ui/d-tree.ts

```diff
@@ -3,19 +3,207 @@
+				Math.floor((height - contentY) / this._itemHeight) + 2 :
+				this._itemOptionsShowable.length
+
+			//get items options are shown in Dpane content
+			const itemOptionsShown = this._itemOptionsShowable.slice(itemIndexMappedStart, itemIndexMappedEnd)
+
+			if (items.length < itemOptionsShown.length) {
+				for (let i = items.length; i < itemOptionsShown.length; i++) {
+					const itemOptions = itemOptionsShown[i]
+					const treeItem = new DTreeItem(itemOptions)
+					content.addChild(treeItem)
+					//listen select item event
+					treeItem.on(UtilPointerEvent.down, (): void => {
+						this.emit("select", treeItem)
+						if (treeItem.isParent()) {
+							this.onToggle(itemOptions.itemPosition)
+						}
+					});
+				}
+			} else if (items.length > itemOptionsShown.length) {
+				for (let i = itemOptionsShown.length; i < items.length; i++) {
+					content.removeChild(items[i])
+					items.length = itemOptionsShown.length
+				}
+			}
+			for (let i = 0; i < items.length; i++) {
+				items[i] = items[i].updateItem(itemOptionsShown[i], this._selectedPosition)
+			}
+		}
+
+		protected onToggle(itemPosition: number[]) {
+			if (this._value instanceof Array) {
+				let items = this._value
+				//access to data of selected item in this.value to update expanded attribute
+				for (let i = 0; i < itemPosition.length - 1; i++) {
+					items = items[itemPosition[i]].children
+				}
+				const item = items[itemPosition[itemPosition.length - 1]]
+				item.expanded = !item.expanded
+
+				//reset data of tree widget
+				this._itemOptions = {}
+				this._itemOptionsShowable.length = 0
+				this._itemY = 0
+				this._content.removeAllListeners()
+				this._content.removeChildren()
+
+				//re-render tree
+				this.mapped([], this._value)
+				this.update()
+				this.registerListeners()
+			}
+		}
+
+		private mapped(parentPosition: number[], items: Array<DTreeItemJsonData>) {
+				for (let i = 0; i < items.length; i++) {
+					const item = items[i];
+					if (item) {
+						//position of item in this._value
+						const itemPosition: number[] = parentPosition.concat([i])
+
+						//calculate padding value of item
+						let paddingValue: number = itemPosition.length * 25
+						const isParent: boolean = item.children && (item.children.length > 0)
+						paddingValue = isParent ? paddingValue - 20 : paddingValue
+						const itemName = item.name ? item.name : ""
+
+						//set default expand status of item is false
+						if (item.expanded == undefined || item.expanded == null) {
+							item.expanded = false
+						}
+
+						let itemOptions : DTreeItemOptions= {
+							itemPosition: itemPosition,
+							text: {
+								value: itemName
+							},
+							y: this._itemY,
+							isParent: isParent,
+							expanded: item.expanded,
+							padding: {
+								left: paddingValue
+							}
+						}
+
+						const parentItemOptions = this._itemOptions[parentPosition.join('-')]
```

Comment:
Same as the above.

Reply:
Should I use "this._itemOptions[parentPosition.join('-')]" directly instead of creating and using "parentItemOptions"?

Reply:
Ditto.

Reply:
I thought we can avoid creating a string instance (parentPosition.join('-')) here.

Reply:
"this._itemOptions" is a dynamic Object, save data with key-value format.
To access the data value in "this._itemOptions" I use  "this._itemOptions[<key>]"
The "key" must be either 'string' or 'number', so I convert "parentPosition" to string by join() function.
For the avoiding "parentPosition.join('-')" I have no idea. Could you give me some suggestions?
Thank you

Reply:
How about something like this?
```
private mapped( parentItemOptions: DItemOptions | null, items: Array<DTreeItemJsonData> ) {
	for (let i = 0; i < items.length; i++) {
		const item = items[i];

		...

		let itemOptions : DTreeItemOptions= { ... };

		...

		if (item && item.children) {
			this.mapped( itemOptions, item.children )
		}
	}
}
```